Restart button for missions

whinis opened the pull request:
This Addresses #4874 and follows on Pull Request #7411
Adds a restart button to the abort menu and end game menu to allow restarting of the mission
Fixes bug in Pull Request #7411 that prevented new replay files from being created for restarted missions and crashes in the resulting single file.

JoesParadiese committed:
restart button added for singleplayer in cnc, ra, d2k

pchote commented on OpenRA.Mods.Common/Widgets/ConfirmationDialogs.cs:
```diff
@@ -56,6 +56,37 @@ public static void CancelPrompt(string title, string text, Action onCancel = nul
 			};
 		}
 
+		public static void PromptAbortMission(World world, string title, string text, Action onAbort, Action onCancel = null, Action closeMenu = null)
```
Do you know why the original PR introduced this duplicate method?  It would be better to generalize `PromptConfirmAction` to cover this use case instead of duplicating code and yaml.

pchote commented on OpenRA.Game/Network/ReplayRecorderConnection.cs:
```diff
@@ -89,10 +89,13 @@ public void Receive(Action<int, byte[]> packetFn)
-					writer.Write(data.Length);
-					writer.Write(data);
-					packetFn(client, data);
+					if (!disposed)
```
I don't think its valid to do anything in here if its been disposed... how about adding an `if (disposed)\nreturn;` at the top instead?

pchote commented on OpenRA.Game/Game.cs:
```diff
@@ -172,7 +172,32 @@ internal static void StartGame(string mapUID, WorldType type)
+			bool record = false;
+			string difficulty = OrderManager.LobbyInfo.GlobalSettings.Difficulty;
+			string gameSpeed = OrderManager.LobbyInfo.GlobalSettings.GameSpeedType;
+			if (OrderManager.Connection is ReplayRecorderConnection)
```
you can simplify this down to `var record = OrderManager.Connection is ReplayRecorderConnection`.  Please also move it to just before the `JoinServer` call where it is used.

pchote commented on OpenRA.Game/Game.cs:
```diff
@@ -172,7 +172,32 @@ internal static void StartGame(string mapUID, WorldType type)
-		public static bool IsHost
+		public static void RestartGame()
+		{
+			string uid = OrderManager.World.Map.Uid;
```
OpenRA code style is to use `var` instead of explicitly typing local variables.

pchote commented on OpenRA.Game/Game.cs:
```diff
@@ -172,7 +172,32 @@ internal static void StartGame(string mapUID, WorldType type)
+				om.IssueOrder(Order.Command("difficulty {0}".F(difficulty)));
+				LobbyInfoChanged -= lobbyReady;
+				om.IssueOrder(Order.Command("state {0}".F(Session.ClientState.Ready)));
+			};
```
Style nit: newline after this please.

pchote commented:
The general approach here looks good to me, although I haven't tested it yet.  A few code nits as noted.

pchote commented on OpenRA.Mods.Common/Widgets/ConfirmationDialogs.cs:
```diff
@@ -15,13 +15,25 @@ namespace OpenRA.Mods.Common.Widgets
+			string confirmText = null,
+			string cancelText = null,
+			World world = null,
+			Action onRestart = null,
```
Please put this with the other Actions.

whinis replied:
If I change the action and strings around all current calls to PromptConfirmAction will require a null before the labels because they do not use restart.

pchote commented on OpenRA.Mods.Common/Widgets/ConfirmationDialogs.cs:
```diff
@@ -15,13 +15,25 @@ namespace OpenRA.Mods.Common.Widgets
+			string cancelText = null,
+			World world = null,
+			Action onRestart = null,
+			string restartText = null)
```
Please put this with the other labels.

pchote commented on OpenRA.Mods.Common/Widgets/ConfirmationDialogs.cs:
```diff
@@ -37,6 +49,23 @@ public static void PromptConfirmAction(string title, string text, Action onConfi
+			var restartButton = prompt.Get<ButtonWidget>("RESTART_BUTTON");
+			if (world != null)
+			{
+				var isVisible = world.IsReplay || world.Map.Visibility == MapVisibility.MissionSelector; // only show for missions and replays
```
This check should be done by the code calling `PromptConfirmAction` instead of hardcoding it here.

pchote replied:
You should then be able to remove the `world` argument.

pchote commented on OpenRA.Mods.Common/Widgets/ConfirmationDialogs.cs:
```diff
@@ -37,6 +49,23 @@ public static void PromptConfirmAction(string title, string text, Action onConfi
+						onRestart();
+
+					Ui.CloseWindow();
+					Game.RunAfterTick(() => Game.RestartGame());
```
`Game.RunAfterTick(Game.RestartGame)`.

pchote commented on OpenRA.Game/Game.cs:
```diff
@@ -172,6 +172,35 @@ internal static void StartGame(string mapUID, WorldType type)
 
+		public static void RestartGame()
+		{
+			if (OrderManager.Connection is ReplayConnection)
```
The disconnect special casing doesn't sit well with me - we don't do this in the normal abort case, so it shouldn't be needed here.  How about something like:

```csharp
var replay = OrderManager.Connection as ReplayConnection;
var replayName = replay != null ? replay.Filename : null;

// Disconnect from the current game
Disconnect();

// Restart the game with the same replay/mission
if (replay != null)
	JoinReplay(i.Filename);
else
{
	// the rest of the code
}
```

whinis commented:
So I attempted to check if this code would work with skirmish missions but there is currently no easy way of syncing the settings or bots again. However as currently pushed it does work for replays (allowing you to start at the beginning of the replay) and for all missions tested so far in mod RA,CNC, and D2k

atlimit8 commented:
Style fix commits should be squashed away.

whinis commented:
Im not really sure how to better squash as with talks to pchote in IRC he said I should split my commits due to changes to PromptConfirmAction. Also some of my work relies on previous work but after these rewrites im not sure how much.

Phrohdoh commented on OpenRA.Game/Game.cs:
```diff
@@ -172,6 +172,41 @@ internal static void StartGame(string mapUID, WorldType type)
+
+			// Restart the game with the same replay/mission
+			if (replay != null)
+			{
```
Unnecessary brackets.

Phrohdoh commented on OpenRA.Game/Network/ReplayRecorderConnection.cs:
```diff
@@ -79,21 +79,23 @@ void StartSavingReplay(byte[] initialContent)
 
 		public void Receive(Action<int, byte[]> packetFn)
 		{
+			if (disposed) // if has been disposed do no continue as writer nolonger exist
```
This comment will need to be fixed.
It should go above the `if` line and grammatically corrected, if we are going to keep it.

abcdefg30 replied:
It should also start with a captial.

Phrohdoh commented on OpenRA.Game/Network/ReplayRecorderConnection.cs:
```diff
@@ -79,21 +79,23 @@ void StartSavingReplay(byte[] initialContent)
 		public void Receive(Action<int, byte[]> packetFn)
 		{
+			if (disposed) // if has been disposed do no continue as writer nolonger exist
+				return;
```
Please add a newline after this.

Phrohdoh commented on OpenRA.Mods.Common/Widgets/ConfirmationDialogs.cs:
```diff
@@ -37,6 +48,22 @@ public static void PromptConfirmAction(string title, string text, Action onConfi
 					onCancel();
 			};
+
+			var restartButton = prompt.Get<ButtonWidget>("RESTART_BUTTON");
```
Please use `GetOrDefault` followed by a null check.
Mods should not be required to add this button.

Phrohdoh commented on OpenRA.Mods.Common/Widgets/Logic/ReplayBrowserLogic.cs:
```diff
@@ -444,16 +443,15 @@ void SetupManagement()
-						"Delete {0} replays?".F(list.Count),
-						() =>
+						title: "Delete all selected replays?",
+						text: "Delete {0} replays?".F(list.Count),
```
`Delete 1 replays?` will look a bit unpolished.

Phrohdoh commented on mods/cnc/chrome/dialogs.yaml:
```diff
@@ -199,7 +206,7 @@ Container@CANCEL_PROMPT:
 			Height: 35
 			Text: Cancel
-
+			
```
Spurious whitespace.

atlimit8 commented:
Sorry for confusing you. I saw that the commit messages say "style fix". Generic "style fix" short log descriptions make it harder to review history, so I will mark them for requested changes to keep the short log understandable. Please replace the short log descriptions with clearer descriptions of what the commits do.

whinis commented:
atlimit8 What would be the best way to handle that? I am not the person who made that commit nor do I have contact with them.

atlimit8 commented:
Do you know how to use `git rebase -i`?

atlimit8 commented:
The commits can be rewritten with a rebase; no need to contact the original author(s).

whinis commented:
That might be better, I squashed the style changes one

whinis commented:
Any update ?

whinis commented:
Further Squashed

Mailaender commented:
Works as promised. :white_check_mark:

pchote commented on OpenRA.Game/Game.cs:
```diff
@@ -172,6 +172,40 @@ internal static void StartGame(string mapUID, WorldType type)
+			{
+				OrderManager om = null;
+				Action lobbyReady = null;
+				var difficulty = globalSettings.Difficulty;
```
Style nit: these can be moved inline with the `Order.Command` lines below

pchote commented on OpenRA.Game/Network/ReplayRecorderConnection.cs:
```diff
@@ -79,21 +79,25 @@ void StartSavingReplay(byte[] initialContent)
 
 		public void Receive(Action<int, byte[]> packetFn)
 		{
+			// If has been disposed do not continue as writer no longer exist
```
Grammar nit: "no longer _exists_"

pchote commented on OpenRA.Mods.Common/Widgets/ConfirmationDialogs.cs:
```diff
@@ -15,13 +15,25 @@ namespace OpenRA.Mods.Common.Widgets
+			string restartText = null)
 		{
 			var prompt = Ui.OpenWindow("CONFIRM_PROMPT");
+			var restartButton = prompt.GetOrNull<ButtonWidget>("RESTART_BUTTON");
```
Could you pull out variables for `confirmButton` and `cancelButton` while you're here too?  That saves two unnecessary widget lookups.

pchote replied:
This is a general purpose third button, so I suggest that we name it `otherButton` and `"OTHER_BUTTON"`

pchote commented on OpenRA.Mods.Common/Widgets/ConfirmationDialogs.cs:
```diff
@@ -37,6 +49,21 @@ public static void PromptConfirmAction(string title, string text, Action onConfi
+					if (onRestart != null)
+						onRestart();
+
+					Ui.CloseWindow();
```
These should be pulled outside this function so that this can be kept as a general purpose button.

pchote commented on OpenRA.Mods.Common/Widgets/ConfirmationDialogs.cs:
```diff
@@ -37,6 +49,21 @@ public static void PromptConfirmAction(string title, string text, Action onConfi
+
+			if (!string.IsNullOrEmpty(restartText) && restartButton != null)
+			{
+				restartButton.IsVisible = () => true;
```
It would make more sense to toggle the visibility (for all three buttons) based on whether the action is null or not.

pchote commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/IngameMenuLogic.cs:
```diff
@@ -88,14 +88,20 @@ public IngameMenuLogic(Widget widget, World world, Action onExit, WorldRenderer
-				ConfirmationDialogs.PromptConfirmAction("Abort Mission", "Leave this game and return to the menu?", onQuit, showMenu);
+				var restartText = "";
+				if (world.IsReplay || world.Map.Visibility == MapVisibility.MissionSelector)
+					restartText = "Restart";
```
Once you've extracted the restart-specific code out of the dialog code and toggle the button visibility based on the action, you can then set the restartAction to be non-null here (basically just assign the block that you pull out of the helper).

pchote commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/IngameMenuLogic.cs:
```diff
@@ -88,14 +88,20 @@ public IngameMenuLogic(Widget widget, World world, Action onExit, WorldRenderer
+					onRestart: closeMenu,
+					confirmText: "Abort",
+					cancelText: "Cancel",
+					restartText: restartText);
```
You can then define `restartText: "Restart"` here.

pchote commented on OpenRA.Game/Game.cs:
```diff
@@ -172,6 +172,38 @@ internal static void StartGame(string mapUID, WorldType type)
+				{
+					om.IssueOrder(Order.Command("gamespeed {0}".F(globalSettings.GameSpeedType)));
+					om.IssueOrder(Order.Command("difficulty {0}".F(globalSettings.Difficulty)));
+					LobbyInfoChanged -= lobbyReady;
```
Style nit: this should go after the IssueOrders, not in the middle(!).

pchote commented on OpenRA.Mods.Common/Widgets/ConfirmationDialogs.cs:
```diff
@@ -15,28 +15,60 @@ namespace OpenRA.Mods.Common.Widgets
+				cancelButton.GetText = () => cancelText;
 
-			prompt.Get<ButtonWidget>("CONFIRM_BUTTON").OnClick = () =>
+			confirmButton.OnClick = () =>
```
You'll now need null checks before using these variables (since `GetOrNull` can return null – `Get` would immediately throw an exception).

pchote commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/IngameMenuLogic.cs:
```diff
@@ -88,14 +88,25 @@ public IngameMenuLogic(Widget widget, World world, Action onExit, WorldRenderer
-				ConfirmationDialogs.PromptConfirmAction("Abort Mission", "Leave this game and return to the menu?", onQuit, showMenu);
+				Action restartAction = null;
+				if (world.IsReplay || world.Map.Visibility == MapVisibility.MissionSelector)
+					restartAction = () =>
```
You've got a hanging block here, so please add braces around it.

whinis commented:
Found a crash actually while testing, if a mission implements https://github.com/OpenRA/OpenRA/blob/bleed/OpenRA.Mods.Common/Scripting/Global/UserInterfaceGlobal.cs then it will crash whenever you attempt to restart the replay. However restarting the actual missions itself is fine.

ScottNZ commented on OpenRA.Game/Game.cs:
```diff
@@ -172,6 +172,38 @@ internal static void StartGame(string mapUID, WorldType type)
 			GC.Collect();
 		}
 
+		public static void RestartGame()
```
I think it's weird to have this Game.RestartGame() in the same file as Game.StartGame() which works completely differently to it, and StartMission() (the real corresponding method to RestartGame()) way out in MissionBrowserLogic.

I'd like to see RestartGame() and StartMission() moved together if possible.

whinis replied:
The plan is for RestartGame to possibly do more in the future but right now it restarts missions and replays. As for moving them together, I'm not exactly sure how I would do that.

ScottNZ replied:
Maybe move https://github.com/OpenRA/OpenRA/blob/bleed/OpenRA.Mods.Common/Widgets/Logic/MissionBrowserLogic.cs#L311-L326 into this file, add a map uid parameter to it, rename it to something more appropriate, and call it from MissionBrowserLogic.

whinis replied:
Actually let me change that and push it

whinis commented:
Problem with GetOrNull has been fixed, a future PR will make the PromptConfirmAction function more generic to allow one message with 1-3 buttons rather than 2 functions.

whinis commented:
Previously mentions crash no longer happens after #10304 was merged

pchote commented on mods/cnc/chrome/dialogs.yaml:
```diff
@@ -156,19 +156,26 @@ Container@CONFIRM_PROMPT:
+			Text: Abort
+		Button@OTHER_BUTTON:
+			Key: r
+			X: 180
```
The UI standard for TD is to group related buttons with 10px clear space between them, with 'forward' moving actions to the right, and 'back' moving actions to the left.  Please move this 30px to the right to fit in with that.

whinis commented:
Is this what you meant pchote? it looks a little odd now

pchote commented:
Yes, that is what I mean.  Dialogs now look fine to me.

One (hopefully) last issue: pressing restart should do the "battle control terminated" notification and fade out instead of immediately jumping to the load screen.

pchote commented:
It would also be good if you could rebase this so that we don't need to clear and re-fetch dependencies when testing.

whinis commented:
Alright, I rebased for sake of testing and made it so the mission will fade before restarting. I noticed an odd bug while doing this in that if missiles are fired while its fading their trails stay on top of the fade.

RoosterDragon commented on OpenRA.Game/Network/ReplayRecorder.cs:
```diff
@@ -69,6 +69,9 @@ void StartSavingReplay(byte[] initialContent)
 		public void Receive(int clientID, byte[] data)
 		{
+			if (disposed)
+				return;
```
Is this needed to prevent a crash? I feel we should fix the caller to not call methods on a disposed object if they're not going to work any more, rather than hiding the problem here.

whinis replied:
So this is to prevent a crash but there isn't a way to check for disposed currently (other than making a public variable we check). Currently this is the only disposed object that crashes and that's because of the writer (once disposed accessing it causes an exception) and my understanding is that currently no other disposed object is checked because of this.

RoosterDragon replied:
What I'm getting at is: we should fix it so nobody calls `Receive` after we've already disposed `ReplayRecorder`. We *should* crash if `ReplayRecorder` is getting new stuff after being disposed, because that likely indicates a bug.

whinis replied:
As we talked about in IRC, I wouldn't be comfortable removing this piece of code until an upstream fix is implemented because otherwise it will crash every restart. Which would mean a broken PR

RoosterDragon replied:
Sorry for not getting to that yet - that's pretty lame of me.

Could you add this comment:  `TODO: This can be removed once NetworkConnection is fixed to dispose properly.`

That'll remind me to get to it eventually, and we can take your PR in the meantime.

whinis commented:
Todo Added and fade code finally implemented. I forgot to push it it seems

whinis commented:
Rebased to be safe

obrakmann commented:
Do you want to update the author on the last commit before merging?

whinis commented:
Will do, should be home soon

obrakmann commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/IngameMenuLogic.cs:
```diff
@@ -88,14 +88,38 @@ public IngameMenuLogic(Widget widget, World world, Action onExit, WorldRenderer
 
 			abortMissionButton.OnClick = () =>
 			{
-				if (world.IsGameOver)
```
I'd prefer it if we could keep this check (in spirit). It would probably need to be a bit smarter and only show when the player has not won, ie. something like
```csharp
if (world.LocalPlayer != null && world.LocalPlayer.WinState != WinState.Won)
   // do the prompt thing
else
{
    onQuit();
    return;
}
```

If there has already been discussion about this, I apologize, I missed it then.

obrakmann commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/IngameMenuLogic.cs:
```diff
@@ -88,14 +88,38 @@ public IngameMenuLogic(Widget widget, World world, Action onExit, WorldRenderer
+					text: "Leave this game and return to the menu?",
+					onConfirm: onQuit,
+					onCancel: showMenu,
+					confirmText: "Abort",
```
Otherwise, if you *don't* add back the check mentioned above, I'd like you to change all instances of "Abort" to "Leave". Otherwise you complete a mission and the dialog will ask you if you want to abort, which would be strange.

Phrohdoh replied:
I agree about the button text changes, "cancel" doesn't tell you what exactly you're canceling.
Please read http://ux.stackexchange.com/a/9960 if you have the time.

whinis replied:
I was just copying over what was mostly there but I do see the value in showing a restart button even if you won the game. What if you wanted to restart the mission to complete a secondary objective you missed? So what would you rather happen since it appears the old behavior was after you win you have the option of staying "cancel" or leaving "abort".

obrakmann commented:
This looks good otherwise, well done. I'll take a final look after you've done the changes.

whinis commented:
obrakmann Is that what you wanted?

Phrohdoh commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/IngameMenuLogic.cs:
```diff
@@ -88,14 +88,48 @@ public IngameMenuLogic(Widget widget, World world, Action onExit, WorldRenderer
+				var cancelText = "Cancel";
+				var otherText = "Restart";
+
+				if (world.LocalPlayer == null || world.LocalPlayer.WinState == WinState.Won)
```
What if you have lost?

Phrohdoh commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/IngameMenuLogic.cs:
```diff
@@ -88,14 +88,48 @@ public IngameMenuLogic(Widget widget, World world, Action onExit, WorldRenderer
+						{
+							if (Game.IsCurrentWorld(world))
+								mpe.Fade(MenuPaletteEffect.EffectType.Black);
+							exitDelay += 40 * mpe.Info.FadeLength;
```
If this `40` is ticks it should probably use the current gamespeed.

whinis replied:
Im using the same code as https://github.com/whinis/OpenRA/blob/bleed/OpenRA.Mods.Common/Widgets/Logic/Ingame/IngameMenuLogic.cs#L50, I'm not sure if its suppose to be 40 ticks but does produce a fading effect that pchote requested.

Phrohdoh replied:
It is definitely ticks and should use gamespeed, but that can be left for another PR because your reasoning for not changing it is valid.

[Reference](https://github.com/OpenRA/OpenRA/blob/da4f04acfb2983038d07baa52d1cc42fcfafcd31/OpenRA.Mods.Common/Traits/PaletteEffects/MenuPaletteEffect.cs#L21-L22).

whinis committed:
Changes to the RestartGame function to make it more streamlined
Removal of PromptAbortAction and inclusion into PromptConfirmAction
Changes to prevent a restart button being required for all mods
ConfirmAction
Addtion of named parameters to PromptConfirmAction
Moved StartGame from MissionBrowserLogic.cs to Game.cs

whinis commented:
That should be a little more clear then, im not an expert on that area but tried to make it a little more context sensitive.

obrakmann commented:
> obrakmann Is that what you wanted?

Not quite. I'd like `abortMissionButton.OnClick` to look like this:

```csharp
			abortMissionButton.OnClick = () =>
			{
				hideMenu = true;
				Action restartAction = null;
				if (world.IsReplay || world.Map.Visibility == MapVisibility.MissionSelector)
				{
					var iop = world.WorldActor.TraitsImplementing<IObjectivesPanel>().FirstOrDefault();
					var exitDelay = iop != null ? iop.ExitDelay : 0;
					restartAction = () =>
					{
						Ui.CloseWindow();
						if (mpe != null)
						{
							if (Game.IsCurrentWorld(world))
								mpe.Fade(MenuPaletteEffect.EffectType.Black);
							exitDelay += 40 * mpe.Info.FadeLength;
						}

						Game.RunAfterDelay(exitDelay, () =>
						{
							Game.RestartGame();
						});
					};
				}

				if (world.LocalPlayer == null || world.LocalPlayer.WinState != WinState.Won &&
					(!world.IsGameOver || world.Map.Visibility == MapVisibility.MissionSelector))
				{
					ConfirmationDialogs.PromptConfirmAction(
						title: "Leave Mission",
						text: "Leave this game and return to the menu?",
						onConfirm: onQuit,
						onCancel: showMenu,
						confirmText: "Leave",
						cancelText: "Stay",
						otherText: "Restart",
						onOther: restartAction);
				}
				else
					onQuit();
			};
```

That will result in the following behaviour:

* Any game, in progress => Show prompt
* Skirmish game, when game is over => Don't show prompt
* Mission, when game is over
    * if the player won => Don't show prompt
    * if the player lost => Show prompt

pchote commented:
Not that `Game.RunAfterDelay(exitDelay, Game.RestartGame)` should work in the above.

obrakmann commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/IngameMenuLogic.cs:
```diff
@@ -88,14 +88,42 @@ public IngameMenuLogic(Widget widget, World world, Action onExit, WorldRenderer
+				hideMenu = true;
+
+				if ((world.LocalPlayer == null || world.LocalPlayer.WinState != WinState.Won) &&
+								   (!world.IsGameOver || world.Map.Visibility == MapVisibility.MissionSelector))
```
`if (world.LocalPlayer == null || (world.LocalPlayer.WinState != WinState.Won && (!world.IsGameOver || world.Map.Visibility == MapVisibility.MissionSelector)))`

Also a few tabs too much on the second line, it seems.

obrakmann commented:
Thanks, looks good now. :+1:

obrakmann commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/35181146a279dc1ee15c2f4ebcf5f7b9a67439b8)
